### ISSUE TITLE
error: Failed to open TCP connection to localhost:8080 (Cannot assign requested address - connect(2) for "localhost" port 8080)

### ISSUE THREAD

**Issue by dwebe003** [onc-healthit/onc-certification-g10-test-kit#77]
error: Failed to open TCP connection to localhost:8080 (Cannot assign requested address - connect(2) for "localhost" port 8080)

I have been tinkering around with docker-compose.yml and some other command line settings, but to no avail, so I thought at this point it might be best to ask here.

On my machine, I have a local web application (built in java) that is running on localhost:8080. It provides the FHIR endpoints that Inferno checks, and separately I have an OAuth microservice running in order to authenticate. 
When I have my inferno docker up and running, and I attempt to run the Standalone Patient App - Full Access test, I come across this error.

I figured this was happening because the Inferno container isn't able to access the localhost that I can on my machine. However, I can't for the life of me figure out how to make that possible. Seems like a network bridge is needed, which I've attempted, but I must be doing something wrong in the context of this repo. 

Not sure if you folks would know how to help in my peculiar situation, but any help is appreciated.

**Comment by dwebe003** [comment]
Some additional info from the container console:

worker_1             | I, [2022-04-08T00:42:59.267681 #1]  INFO -- : Initializing client with http://localhost:8080/api/fhir/r4

worker_1             | I, [2022-04-08T00:42:59.267820 #1]  INFO -- : Configuring the client to use no authentication.

worker_1             | I, [2022-04-08T00:42:59.269268 #1]  INFO -- : GETTING: http://localhost:8080/api/fhir/r4/metadata

worker_1             | E, [2022-04-08T00:42:59.271611 #1] ERROR -- : GET - Request: http://localhost:8080/api/fhir/r4/metadata failed! No response from server: Failed to open TCP connection to localhost:8080 (Cannot assign requested address - connect(2) for "localhost" port 8080)

worker_1             | E, [2022-04-08T00:42:59.272388 #1] ERROR -- : /usr/local/lib/ruby/2.7.0/net/http.rb:960:in `initialize': Failed to open TCP connection to localhost:8080 (Cannot assign requested address - connect(2) for "localhost" port 8080) (Errno::EADDRNOTAVAIL)

**Comment by dwebe003** [comment]
UPDATE: Soon after posting this I was able to figure out that my API URL just needs to contain my shared network IP rather than localhost. All good now. Able to authenticate successfully and retrieve an access_token. 

However, when I move on over to the Single Patient API Test, the `Submit` button is disabled/greyed-out on the Run All Tests modal. All the necessary info seems to be there: the FHIR endpoint, the SMART App Launch Credentials, and the Patient ID. Any idea why this could be?

**Comment by arscan** [comment]
Glad you were able to figure out how to connect to a service on your same machine.  We should add a section in the README or similar with instructions on how to do this.  Using your machine's IP address is one option, another is to use `host.docker.internal`, which is a host that docker sets up for your containers to connect to the host machine. That works for me, though I've heard others have had issues with that approach.

Hmmm, we should add some kind of message about what is missing, if it isn't obvious.  So you already ran one of the launch tests to get a valid authorization token, then clicked on 'Single Patient API' and then 'Run Tests'?  Could you provide a screenshot of the modal?

**Comment by dwebe003** [comment]
I'm glad too, it was rough for a minute there. 

Yep that's exactly what I did. Here's a screenshot. I'm wondering if it is due to `Patient ID from SMART App Launch` field being empty? 
![Screen Shot 2022-04-07 at 6 59 41 PM](https://user-images.githubusercontent.com/6394944/162348595-45c16bce-7007-4c93-a825-c024596e8173.png)

**Comment by arscan** [comment]
Hmmm, I'm not sure why you are getting Patient IDs listed twice.  I take it you are running off a clone of `main`?  Did you get a patient context in your launch, or did you type in the "Patient IDs (required)" field in the bottom?

At the bottom of the page, what version of inferno-core are you using?

![Screen Shot 2022-04-07 at 10 21 37 PM](https://user-images.githubusercontent.com/412901/162350585-58fff5ff-6596-48ed-a920-b704f9260660.png)

**Comment by dwebe003** [comment]
Running off a clone, yep. Cloned the day before yesterday.
No patient context, I typed that patient ID myself. 
Same version. 0.2.0 

Perhaps I'll try a fresh clone tomorrow morning and see what happens.

**Comment by arscan** [comment]
Thanks, we'll take a closer look tomorrow, and may have follow-up questions.  I think this issue might be resolved if the previous launch you did provided a patient in context, but there are other strange things going on here I'd like to fix ASAP.  Thank you for the report!

**Comment by dwebe003** [comment]
@arscan I was able to execute the Single Patient API tests just now. I realized it was because my OAuth response doesn't contain the patient ID that I authenticated with, which explains why my "Patient ID from SMART App Launch" field was empty. 
My team and I will have to go in and add that to our OAuth response, but for now I alleviated this by opening single_patient_api_group.rb and changing that field's `locked` property from true to false. So now I could manually enter the ID.
![Screen Shot 2022-04-08 at 9 29 39 AM](https://user-images.githubusercontent.com/6394944/162483900-9582cd67-5927-414e-b5a6-0baa2dec8128.png)
